crates/extensions/c8y_http_proxy/src/actor.rs

```diff
@@ -162,149 +167,241 @@ impl C8YHttpProxyActor {
-        &mut self,
-        child_device_id: String,
-    ) -> Result<String, C8YRestError> {
-        if let Some(c8y_internal_id) = self.child_devices.get(&child_device_id) {
-            Ok(c8y_internal_id.clone())
-        } else {
-            let c8y_internal_id = self.try_get_internal_id(Some(&child_device_id)).await?;
-            self.child_devices
-                .insert(child_device_id, c8y_internal_id.clone());
-            Ok(c8y_internal_id)
+    async fn try_get_internal_id(&mut self, device_id: String) -> Result<String, C8YRestError> {
+        let url_get_id: String = self.end_point.get_url_for_internal_id(device_id);
+        if self.end_point.token.is_none() {
+            self.get_fresh_token().await?;
         }
-    }
-
-    async fn try_get_internal_id(
-        &mut self,
-        device_id: Option<&str>,
-    ) -> Result<String, C8YRestError> {
-        let url_get_id = self.end_point.get_url_for_get_id(device_id);
+        let request = HttpRequestBuilder::get(&url_get_id)
+            .bearer_auth(self.end_point.token.clone().unwrap_or_default())
+            .build()?;
+        let res = match self.peers.http.await_response(request).await? {
+            Ok(response) => match response.status() {
+                StatusCode::OK => Ok(Ok(response)),
+                StatusCode::UNAUTHORIZED | StatusCode::FORBIDDEN => {
+                    self.get_fresh_token().await?;
+                    let request = HttpRequestBuilder::get(&url_get_id)
+                        .bearer_auth(self.end_point.token.clone().unwrap_or_default())
+                        .build()?;
+                    Ok(self.peers.http.await_response(request).await?)
+                }
+                code => Err(C8YRestError::FromHttpError(
+                    tedge_http_ext::HttpError::HttpStatusError(code),
+                )),
+            },
 
-        let request_internal_id = HttpRequestBuilder::get(url_get_id);
-        let res = self.execute(request_internal_id).await?;
-        let res = res.error_for_status()?;
+            Err(e) => Err(C8YRestError::FromHttpError(e)),
+        };
+        let res = res?.error_for_status()?;
 
         let internal_id_response: InternalIdResponse = res.json().await?;
         let internal_id = internal_id_response.id();
+
         Ok(internal_id)
     }
 
     async fn execute(
         &mut self,
-        request_builder: HttpRequestBuilder,
+        device_id: String,
+        build_request: impl Fn(&C8yEndPoint) -> Result<HttpRequestBuilder, C8YRestError>,
     ) -> Result<HttpResult, C8YRestError> {
-        // Get a JWT token to authenticate the device
-        let request_builder = if let Ok(token) = self.peers.jwt.await_response(()).await? {
-            request_builder.bearer_auth(token)
-        } else {
-            return Err(C8YRestError::CustomError("JWT token not available".into()));
-        };
+        let device_id = self.end_point.get_device_id(device_id);
```

Comment:
That looks really weird. Deriving the `device_id` from the `device_id`. I'm assuming that this is happening because of that special handling of "main" literal. If possible, I'd avoid passing that special char around.

Reply:
refactored